### PR TITLE
Fix ARM64 unsigned div by const perf regression

### LINKED ISSUE #54166: [Perf] Regressions in Integer Formatting
### Run Information

Architecture | arm64
-- | --
OS | Windows 10.0.19041
Baseline | [4c7f4ee74d7a9e5bcc73b552dd020df02b039c8a](https://github.com/dotnet/runtime/commit/4c7f4ee74d7a9e5bcc73b552dd020df02b039c8a)
Compare | [8fb2eb53f75971ce492f14bebb93ed7a236bcd6e](https://github.com/dotnet/runtime/commit/8fb2eb53f75971ce492f14bebb93ed7a236bcd6e)
Diff | [Diff](https://github.com/dotnet/runtime/compare/4c7f4ee74d7a9e5bcc73b552dd020df02b039c8a...8fb2eb53f75971ce492f14bebb93ed7a236bcd6e)

### Regressions in System.Tests.Perf_Int32

Benchmark | Baseline | Test | Test/Base | Test Quality | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
[ToString - Duration of single invocation](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_arm64_Windows 10.0.19041/System.Tests.Perf_Int32.ToString(value%3a%202147483647).html>) | 24.70 ns | 34.44 ns | 1.39 | 0.21 | | |
[TryFormat - Duration of single invocation](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_arm64_Windows 10.0.19041/System.Tests.Perf_Int32.TryFormat(value%3a%202147483647).html>) | 19.39 ns | 21.12 ns | 1.09 | 0.17 | | |

graph
graph
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/05_25_2021/refs/heads/main_arm64_Windows%2010.0.19041_Regression/System.Tests.Perf_Int32.html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
py .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'System.Tests.Perf_Int32*'
```
<details>

### Payloads

[Baseline](<https://helixdi107v0xdeko0k025g8.blob.core.windows.net/helix-job-09b0e867-d918-493d-8f94-1ed57e7c008ca6480ee76424784ac/d17f1bb6-cf7d-43e9-82f4-d7b958031de9.zip?sv=2019-07-07&se=2021-06-17T14%3A13%3A59Z&sr=c&sp=rl&sig=4y%2FfMT1VANdqOiV7YklImmCgyUNsLeW1vgMV5YQWdCo%3D>)
[Compare](<https://helixdi107v0xdeko0k025g8.blob.core.windows.net/helix-job-9a10712e-55d4-4710-8cc4-a7764f27676722b004842d841ee99/e20c8586-25be-495a-9117-95dd22f83a6e.zip?sv=2019-07-07&se=2021-06-17T20%3A36%3A56Z&sr=c&sp=rl&sig=QKoTBO7ix54nB3Ag0uJir2PLiqWfCxIxMffTt5tLT4w%3D>)

#### System.Tests.Perf_Int32.ToString(value: 2147483647)

```log

```
#### System.Tests.Perf_Int32.TryFormat(value: 2147483647)

```log

```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

### Run Information

Architecture | arm64
-- | --
OS | Windows 10.0.19041
Baseline | [4c7f4ee74d7a9e5bcc73b552dd020df02b039c8a](https://github.com/dotnet/runtime/commit/4c7f4ee74d7a9e5bcc73b552dd020df02b039c8a)
Compare | [8fb2eb53f75971ce492f14bebb93ed7a236bcd6e](https://github.com/dotnet/runtime/commit/8fb2eb53f75971ce492f14bebb93ed7a236bcd6e)
Diff | [Diff](https://github.com/dotnet/runtime/compare/4c7f4ee74d7a9e5bcc73b552dd020df02b039c8a...8fb2eb53f75971ce492f14bebb93ed7a236bcd6e)

### Regressions in System.Tests.Perf_UInt64

Benchmark | Baseline | Test | Test/Base | Test Quality | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
[TryFormat - Duration of single invocation](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_arm64_Windows 10.0.19041/System.Tests.Perf_UInt64.TryFormat(value%3a%2012345).html>) | 10.78 ns | 12.11 ns | 1.12 | 0.13 | | |

graph
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/05_25_2021/refs/heads/main_arm64_Windows%2010.0.19041_Regression/System.Tests.Perf_UInt64.html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
py .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'System.Tests.Perf_UInt64*'
```
<details>

### Payloads

[Baseline](<https://helixdi107v0xdeko0k025g8.blob.core.windows.net/helix-job-09b0e867-d918-493d-8f94-1ed57e7c008ca6480ee76424784ac/d17f1bb6-cf7d-43e9-82f4-d7b958031de9.zip?sv=2019-07-07&se=2021-06-17T14%3A13%3A59Z&sr=c&sp=rl&sig=4y%2FfMT1VANdqOiV7YklImmCgyUNsLeW1vgMV5YQWdCo%3D>)
[Compare](<https://helixdi107v0xdeko0k025g8.blob.core.windows.net/helix-job-9a10712e-55d4-4710-8cc4-a7764f27676722b004842d841ee99/e20c8586-25be-495a-9117-95dd22f83a6e.zip?sv=2019-07-07&se=2021-06-17T20%3A36%3A56Z&sr=c&sp=rl&sig=QKoTBO7ix54nB3Ag0uJir2PLiqWfCxIxMffTt5tLT4w%3D>)

#### System.Tests.Perf_UInt64.TryFormat(value: 12345)

```log

```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

### Run Information

Architecture | arm64
-- | --
OS | Windows 10.0.19041
Baseline | [4c7f4ee74d7a9e5bcc73b552dd020df02b039c8a](https://github.com/dotnet/runtime/commit/4c7f4ee74d7a9e5bcc73b552dd020df02b039c8a)
Compare | [8fb2eb53f75971ce492f14bebb93ed7a236bcd6e](https://github.com/dotnet/runtime/commit/8fb2eb53f75971ce492f14bebb93ed7a236bcd6e)
Diff | [Diff](https://github.com/dotnet/runtime/compare/4c7f4ee74d7a9e5bcc73b552dd020df02b039c8a...8fb2eb53f75971ce492f14bebb93ed7a236bcd6e)

### Regressions in System.Tests.Perf_Version

Benchmark | Baseline | Test | Test/Base | Test Quality | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
[TryFormatL - Duration of single invocation](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_arm64_Windows 10.0.19041/System.Tests.Perf_Version.TryFormatL.html>) | 96.02 ns | 103.50 ns | 1.08 | 0.15 | | |

graph
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/05_25_2021/refs/heads/main_arm64_Windows%2010.0.19041_Regression/System.Tests.Perf_Version.html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
py .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'System.Tests.Perf_Version*'
```
<details>

### Payloads

[Baseline](<https://helixdi107v0xdeko0k025g8.blob.core.windows.net/helix-job-09b0e867-d918-493d-8f94-1ed57e7c008ca6480ee76424784ac/d17f1bb6-cf7d-43e9-82f4-d7b958031de9.zip?sv=2019-07-07&se=2021-06-17T14%3A13%3A59Z&sr=c&sp=rl&sig=4y%2FfMT1VANdqOiV7YklImmCgyUNsLeW1vgMV5YQWdCo%3D>)
[Compare](<https://helixdi107v0xdeko0k025g8.blob.core.windows.net/helix-job-9a10712e-55d4-4710-8cc4-a7764f27676722b004842d841ee99/e20c8586-25be-495a-9117-95dd22f83a6e.zip?sv=2019-07-07&se=2021-06-17T20%3A36%3A56Z&sr=c&sp=rl&sig=QKoTBO7ix54nB3Ag0uJir2PLiqWfCxIxMffTt5tLT4w%3D>)

#### System.Tests.Perf_Version.TryFormatL

```log

```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

### Run Information

Architecture | arm64
-- | --
OS | Windows 10.0.19041
Baseline | [28e63279342bd2f6ca43442d864f487613a53bc9](https://github.com/dotnet/runtime/commit/28e63279342bd2f6ca43442d864f487613a53bc9)
Compare | [d49bcbe0441f5c954cddcbe28a222eb34917bcaf](https://github.com/dotnet/runtime/commit/d49bcbe0441f5c954cddcbe28a222eb34917bcaf)
Diff | [Diff](https://github.com/dotnet/runtime/compare/28e63279342bd2f6ca43442d864f487613a53bc9...d49bcbe0441f5c954cddcbe28a222eb34917bcaf)

### Regressions in System.Tests.Perf_UInt32

Benchmark | Baseline | Test | Test/Base | Test Quality | Baseline IR | Compare IR | IR Ratio | Baseline ETL | Compare ETL
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
[ToString - Duration of single invocation](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_arm64_Windows 10.0.19041/System.Tests.Perf_UInt32.ToString(value%3a%204294967295).html>) | 22.92 ns | 25.28 ns | 1.10 | 0.24 | | |
[TryFormat - Duration of single invocation](<https://pvscmdupload.blob.core.windows.net/reports/allTestHistory/refs/heads/main_arm64_Windows 10.0.19041/System.Tests.Perf_UInt32.TryFormat(value%3a%204294967295).html>) | 17.54 ns | 20.77 ns | 1.18 | 0.01 | | |

graph
graph
[Historical Data in Reporting System](<https://pvscmdupload.blob.core.windows.net/autofilereport/autofilereports/05_25_2021/refs/heads/main_arm64_Windows%2010.0.19041_Regression/System.Tests.Perf_UInt32.html>)

### Repro

```cmd
git clone https://github.com/dotnet/performance.git
py .\performance\scripts\benchmarks_ci.py -f netcoreapp5.0 --filter 'System.Tests.Perf_UInt32*'
```
<details>

### Payloads

[Baseline](<https://helixdi8s23ayyeko0k025g8.blob.core.windows.net/helix-job-f2d20920-0bc0-4bb9-9d1d-8d34eb6d47c6c216c972f62465292/99063571-1fe1-4096-8569-59b2e86938a4.zip?sv=2019-07-07&se=2021-06-18T18%3A28%3A15Z&sr=c&sp=rl&sig=CuaIYn6sp0EYxmgqo1u%2FF5V2IjlpjtZgzufOov%2B2mm0%3D>)
[Compare](<https://helixdi8s23ayyeko0k025g8.blob.core.windows.net/helix-job-f7b88df2-2a36-4dbd-a714-925cb640fe3c2d7e9db63214e9fbe/bc83e467-c725-44b8-a0ae-9c7b252f25ee.zip?sv=2019-07-07&se=2021-06-19T01%3A07%3A40Z&sr=c&sp=rl&sig=jgPCNxUQFidzmxgWf%2BJL4S1pK8dKXjvRl2T99A2SlQw%3D>)

#### System.Tests.Perf_UInt32.ToString(value: 4294967295)

```log

```
#### System.Tests.Perf_UInt32.TryFormat(value: 4294967295)

```log

```

### Docs

[Profiling workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/profiling-workflow-dotnet-runtime.md)
[Benchmarking workflow for dotnet/runtime repository](https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md)

</details>

### PR DESCRIPTION
**Pull request by pentp**
This fixes #54166. I added a new `GT_MULWIDE` node for this, but then later realized that we should actually try to fix the `GT_MUL` arm64 codegen for such cases where both operands are 32-bit as a general solution that would benefit for example `System.Decimal` also (it does a lot of such multiplications). What do you think @AndyAyersMS?

Diff: https://www.diffchecker.com/9mkVf4se
Diff with baseline (no #52893): https://www.diffchecker.com/x2SRZJvW

### CODE REVIEW

File: src/coreclr/jit/codegenarm64.cpp [hunk 1]
```diff
@@ -1817,6 +1817,26 @@ void CodeGen::genCodeForMulHi(GenTreeOp* treeNode)
     genProduceReg(treeNode);
 }
 
+// Generate code to get the 64-bits of a 32*32 bit multiplication result
+void CodeGen::genCodeForMulWide(GenTreeOp* treeNode)
```
Comment (SingleAccretion) [inline src/coreclr/jit/codegenarm64.cpp:1821]: Isn't this a `GT_MUL_LONG`? We only use for 32 bit targets now, but it should be straightforward to use it on ARM64 too.

**Comment by AndyAyersMS** [comment]
cc @dotnet/jit-contrib.

Can somebody more familiar with arm64 codegen PTAL. It would be nice to verify this fixes the regression, so likely one of us will need to be on tap to run the benchmarks. Anyone want to sign up to shepherd this fix?

> we should actually try to fix the GT_MUL arm64 codegen...

I like the sound of this. Is it something you can work up reasonably quickly so we can assess how complicated it would be and whether it looks like a more attractive alternative?

**Comment by JulieLeeMSFT** [comment]
@echesakovMSFT PTAL.

**Comment by echesakovMSFT** [comment]
I agree that having `GT_MUL` generating: 
* `smull` for `s32 x s32 -> s64`
* `umull` for `u32 x u32 -> u64`
* `mul` for `i32 x i32 -> i32` and `i64 x i64 -> i64`

would be better that introducing an IR node that only gonna be used for a few cases like this one.
In fact, I opened issue #47490 that targets such work.

I will run the affected benchmarks on Arm64 device on Monday and post results here.

**Comment by EgorBo** [comment]
Heh, I had a prototype for #47490 where I also just removed casts and changed types like this:
```
* MUL TYP_LONG
\ * X, TYP_INT
  * Y, TYP_INT
```
and recognized umull/smull in the codegen for it

**Comment by SingleAccretion** [comment]
We already have all the necessary logic to recognize such trees [in morph](https://github.com/dotnet/runtime/blob/b2fc310998f1d25ed01102a47366f89b5c1c9e9c/src/coreclr/jit/morph.cpp#L14663), but it may need some refactoring to be usable in lowering (I do not think we want it in morph for ARM64). However, if you want the easy way out, there is a "dumbed down" debug-only (but still correct) version that will work on LIR as-is [here](https://github.com/dotnet/runtime/blob/b2fc310998f1d25ed01102a47366f89b5c1c9e9c/src/coreclr/jit/gentree.cpp#L2439). And we already have an oper that represents `32x32 -> 64` multiplication `GT_MUL_LONG`. It would seem to me that the only missing piece is codegen support, which this PR already adds.

**Comment by pentp** [comment]
Closing this one as #57400 is better.

It includes the codegen change for umull/smull, but I don't know what's the correct place for removing casts (optNarrowTree in optimizer.cpp?), so I'll leave it to you @EgorBo as you already have a prototype.

**Pull request closed without merging**